### PR TITLE
[`distributed`] decoupled tp_plan and ep_plan

### PR DESCRIPTION
**Pull request by 3outeille**
[CPU CI](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=48859) [GPU run-slow](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=48859)

# What does this PR do?

Today EP is not its own parallelism axis: it's TP's axis with a different plan swapped in (we use the `_ep_plan` instead of `_tp_plan` when `enable_expert_parallel=True`). Because EP reuses the tp mesh dim, `ep_size == tp_size` which implies that the 2 divisibility constraints collide:

- TP needs `num_kv_heads % tp_size == 0` (`ColwiseParallel` on k/v proj)
- EP needs `num_experts % ep_size == 0` (`GroupedGemmParallel / RouterParallel`)

This PR lifts this constraint now by decoupling TP and EP plan. The follow up PR https://github.com/huggingface/transformers/pull/48865 will be responsible for scaling `tp_size` and `ep_size` independently (but as for now, we default to the EP masked all-reduce)

### CODE REVIEW

File: src/transformers/distributed/mixin.py [hunk 1]
```diff
@@ -175,29 +185,33 @@ def maybe_distribute_model(
...
+        if tp_plan:
+            model = apply_tensor_parallelism(model, mesh_manager.get_mesh("tp"), tp_plan)
+        if ep_plan:
+            # Because we do masked all-reduce EP, we need to shard the experts on TP mesh (tp_size = ep_size).
+            # We can't use ep mesh because it belongs to the expert mesh (efsdp for dispatch path)
+            # which is different from the dense mesh (fsdp).
+            model = apply_masked_expert_parallelism(model, mesh_manager.get_mesh("tp"), ep_plan)
+
+        if distributed_config.fsdp_size > 1:
+            model = apply_fully_sharded_data_parallelism(model, mesh_manager.get_mesh("fsdp"))
```
Comment (ArthurZucker) [inline src/transformers/distributed/mixin.py:212]: would be better, more explicit if instead you did not have the mesh manager, and were directly saying here we need pp, dp, here pp, tp, here pp, dp, ep etc

File: src/transformers/distributed/tensor_parallel.py [hunk 2]
```diff
@@ -792,32 +801,84 @@ class ParallelInterface(GeneralInterface):
...
+) -> tuple[dict[str, str], dict[str, str]]:
+    """Merge the `DistributedConfig` overrides into the model's plans and split them between TP and EP.
 
-    _validate_tp_plan_styles(model.tp_plan)
+    Returns the TP plan to apply to the dense modules and the EP plan to apply to the experts. Each plan is empty
+    when its parallel size is 1. EP owns every module it names, so TP rules for those modules and their children
+    are dropped: expert weights are sharded once, by the EP plan.
+    """
+    # Reject invalid paths before merging, e.g. "layers.*" when the model uses "model.layers.*".
+    model_names = {name for name, _ in model.named_modules()} | {name for name, _ in model.named_parameters()}
```
Comment (ArthurZucker) [inline src/transformers/distributed/tensor_parallel.py]: i'd be careful on this one, we re-compute it over and over and over again. Check if we don't have this stored somewhere (left or right term). model names is a bit weird TBH as variable name

File: src/transformers/distributed/tensor_parallel.py [hunk 3]
```diff
@@ -792,32 +801,84 @@ class ParallelInterface(GeneralInterface):
...
+    if isinstance(distributed_config.tp_plan, dict):
+        valid_names = model_names | set(model.tp_plan)
+        for pattern in distributed_config.tp_plan:
+            if not any(fnmatchcase(name, pattern) for name in valid_names):
```
Comment (ArthurZucker) [inline src/transformers/distributed/tensor_parallel.py]: mmm why are we not using any regex? like you compile one, pipe regex pouf

File: src/transformers/distributed/tensor_parallel.py [hunk 4]
```diff
@@ -792,32 +801,84 @@ class ParallelInterface(GeneralInterface):
...
+                raise ValueError(
+                    f"The `tp_plan` pattern {pattern!r} does not match any module, parameter, "
+                    f"or existing plan entry in {type(model).__name__}. "
+                    "Check the full path, including any 'model.' prefix."
+                )
+
+    if isinstance(distributed_config.ep_plan, dict):
+        valid_names = model_names | set(model.ep_plan)
+        for pattern in distributed_config.ep_plan:
+            if not any(fnmatchcase(name, pattern) for name in valid_names):
```
Comment (ArthurZucker) [inline src/transformers/distributed/tensor_parallel.py]: you can check both at the same time with a good regex

File: src/transformers/distributed/tensor_parallel.py [hunk 5]
```diff
@@ -829,6 +890,15 @@ def apply_tensor_parallelism(model, tp_mesh):
     return model
 
 
+def apply_masked_expert_parallelism(model: nn.Module, tp_mesh: DeviceMesh, ep_plan: dict[str, str]):
```
Comment (ArthurZucker) [inline src/transformers/distributed/tensor_parallel.py]: do we even need that funct?

File: src/transformers/distributed/mixin.py (hunk 1 above)
Comment (3outeille) [reply src/transformers/distributed/mixin.py:212]: that means moving around the `expert_mesh` and `dense_mesh`as func arguments which is annoying.

File: src/transformers/distributed/tensor_parallel.py (hunk 5 above)
Comment (3outeille) [reply src/transformers/distributed/tensor_parallel.py]: we dont need it, will remove

File: src/transformers/distributed/tensor_parallel.py (hunk 2 above)
Comment (3outeille) [reply src/transformers/distributed/tensor_parallel.py]: I checked and it's the first time we compute those terms. Plus, we are calling `resolve_parallel_plans` just one time so that's fine
